projects/com.oracle.truffle.llvm.parser.bc.impl/src/com/oracle/truffle/llvm/parser/bc/impl/LLVMBitcodeVisitor.java

```diff
@@ -172,36 +172,103 @@ public static LLVMParserResult getMain(Source source, LLVMContext context) {
     }
 
-    private LLVMExpressionNode createFunction(FunctionDefinition method) {
-        String name = method.getName();
+    private LLVMExpressionNode createFunction(FunctionDefinition method, LLVMLifetimeAnalysis lifetimes) {
+        String functionName = method.getName();
 
         LLVMBitcodeFunctionVisitor visitor = new LLVMBitcodeFunctionVisitor(
                         this,
-                        frames.getDescriptor(name),
-                        frames.getSlots(name),
-                        labels.labels(name),
-                        phis.getPhiMap(name),
+                        stack.getFrame(functionName),
+                        labels.labels(functionName),
+                        phis.getPhiMap(functionName),
                         method.getParameters().size());
 
         method.accept(visitor);
 
-        LLVMBasicBlockNode[] basicBlocks = visitor.getBlocks();
+        final int[] basicBlockIndices = new int[method.getBlockCount()];
+        for (int i = 0; i < method.getBlockCount(); i++) {
+            basicBlockIndices[i] = i;
+        }
+        final LLVMStackFrameNuller[][] slotNullerBeginNodes = getSlotNuller(method, lifetimes.getNullableBefore());
+        final LLVMStackFrameNuller[][] slotNullerAfterNodes = getSlotNuller(method, lifetimes.getNullableAfter());
+        return LLVMBlockFactory.createFunctionBlock(visitor.getReturnSlot(), visitor.getBlocks(), slotNullerBeginNodes, slotNullerAfterNodes);
+    }
+
+    private static LLVMStackFrameNuller[][] getSlotNuller(FunctionDefinition method, Map<InstructionBlock, FrameSlot[]> slots) {
+        final LLVMStackFrameNuller[][] indexToSlotNuller = new LLVMStackFrameNuller[method.getBlockCount()][];
+        for (int j = 0; j < method.getBlockCount(); j++) {
+            final InstructionBlock block = method.getBlock(j);
+            final FrameSlot[] deadSlots = slots.get(block);
+            if (deadSlots != null) {
+                LLVMParserAsserts.assertNoNullElement(deadSlots);
+                indexToSlotNuller[j] = getSlotNullerNode(deadSlots);
+            }
+        }
+        return indexToSlotNuller;
+    }
+
+    private static LLVMStackFrameNuller[] getSlotNullerNode(FrameSlot[] deadSlots) {
+        if (deadSlots == null) {
+            return new LLVMStackFrameNuller[0];
+        }
+        final LLVMStackFrameNuller[] nullers = new LLVMStackFrameNuller[deadSlots.length];
+        int i = 0;
+        for (FrameSlot slot : deadSlots) {
+            nullers[i++] = getNullerNode(slot);
+        }
+        LLVMParserAsserts.assertNoNullElement(nullers);
+        return nullers;
+    }
+
+    private static LLVMStackFrameNuller getNullerNode(FrameSlot slot) {
+        final String identifier = (String) slot.getIdentifier();
+        return createFrameNuller(identifier, slot);
+    }
+
+    private static LLVMStackFrameNuller createFrameNuller(String identifier, FrameSlot slot) {
```

Comment:
#504 introduces the changes necessary to use a facade here, I did not want to duplicate @pointhi's code, so for now I just copied the code from the facade here.